rtt/base/DataObjectLockFree.hpp

```diff
@@ -249,38 +254,55 @@ namespace RTT
          */
         virtual bool Set( param_t push )
         {
-            /**
-             * This method can not be called concurrently (only one
-             * producer). With a minimum of 3 buffers, if the
-             * write_ptr+1 field is not occupied, it will remain so
-             * because the read_ptr is at write_ptr-1 (and can
-             * not increment the counter on write_ptr+1). Hence, no
-             * locking is needed.
-             */
-
             if (!initialized) {
                 log(Error) << "You set a lock-free data object of type " << internal::DataSourceTypeInfo<T>::getType() << " without initializing it with a data sample. "
                            << "This might not be real-time safe." << endlog();
                 data_sample(value_t(), true);
             }
 
-            // writeout in any case
-            PtrType writing = write_ptr;
+            PtrType writing = write_ptr;            // copy buffer location
+            if (!oro_atomic_inc_and_test(&writing->write_lock)) {
+                // abort, another thread already successfully locked this buffer element
+                oro_atomic_dec(&writing->write_lock);
+                return false;
+            }
+
+            // Additional check that resolves the following race condition:
+            //  - writer A copies the write_ptr and acquires the lock writing->write_lock
+            //  - writer B copies the write_ptr
+            //  - writer A continues to update and increments the write_ptr
+            //  - writer A releases the lock writing->write_lock
+            //  - writer B acquires the lock successfully, but for the same buffer element that already
+            //    has been written by writer A and can potentially be accessed by readers now!
+            if ( writing != write_ptr ) {
+                // abort, another thread already updated the write_ptr, which could imply that read_ptr == writing now
+                oro_atomic_dec(&writing->write_lock);
+                return false;
+            }
+            // from here on we are sure that 'writing'
+            // is a valid buffer to write to and we
+            // have exclusive access
+
+            // copy sample
             writing->data = push;
             writing->status = NewData;
 
             // if next field is occupied (by read_ptr or counter),
             // go to next and check again...
-            while ( oro_atomic_read( &write_ptr->next->counter ) != 0 || write_ptr->next == read_ptr )
+            while ( oro_atomic_read( &write_ptr->next->read_counter ) != 0 ||
```

Comment:
As discussed, this loop should probably not directly manipulate the write_ptr, to prevent preemption in the middle by a writer that would subsequently write to a buffer that is (potentially) being read.

Reply:
If all other writers drop out early, like it is now, there is no problem because it is guaranteed that no other thread enters that section concurrently.

Reply:
When you get preempted in the middle of this loop, stuff could go wrong, no?

Reply:
I don't see why. If only one thread is executing that finite loop, and no other thread ever waits for its completion, there is no problem. Or do I miss something?